enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,375 @@
+ any given call.  As long as the data is there by the time the final provisioning occurs, we are good.
+
+The [secrets store csi plugin](https://github.com/kubernetes-sigs/secrets-store-csi-driver) as very similar to hostpath
+in that 
+- the CSI spec implementation points were easy to find
+- there were golang file system level calls during volume provisioning, i.e. `NodePublishVolume`, like `ioutil.ReadDir`
+- and use of the k8s util mount API's
+- the biggest difference is that it mounts in the `SecretProviderClass` implementation and [calls the provided binary](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/pkg/secrets-store/nodeserver.go#L196-L232)
+to prime the storage as part of provisioning the volume.
+- the azure and hashicorp examples did not work for me on OpenShift because of permission related issues.
+- but **IF** we chose to go down the `SecretProviderClass` path, we would provide our own implementation that would not
+suffer from those permissions problems.
+- See open questions around using this repo as is.
+
+### Risks and Mitigations
+
+While this proposal settles on installing via CVO, that decision has ramifications whose discussion are noted above.
+It is at least conceivable the need to revisit that decision may arise when implementation starts.
+
+Getting the flow and various stages from the CSI specification understood and correctly utilized so that we minimize
+the amount of times the Secret is read and stored on the local filesystem of each host/node is important.  We want
+to minimize local file system writes to when the Secret actually changes, and not when we provision the volume to 
+a new Pod. 
+
+
+## Design Details
+
+### Test Plan
+
+Verification of this proposal should be containable around e2e's that automate various forms of [the basic Hostpath
+driver confirmation](https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/docs/deploy-1.17-and-later.md#confirm-hostpath-driver-works), where that
+confirmation is augmented to consider:
+
+- the multiple replicas of the `DaemonSet`
+- create/update/delete of the K8S Secret(s) designated for host/node injection are properly reflected on the volume
+
+Also, there is already some e2e's around the existing csi-hostpath example.  The ginkgo test focus  `[sig-storage] CSI Volumes [Driver: csi-hostpath]`
+shows up in several of the base e2e suites.
+
+Some links related to that testing:
+- https://github.com/openshift/origin/blob/master/cmd/openshift-tests/csi.go
+- https://github.com/openshift/origin/tree/master/test/extended/testdata/csi
+- https://github.com/openshift/origin/tree/master/vendor/k8s.io/kubernetes/test/e2e/storage/external
+
+### Graduation Criteria
+
+This should be introduced directly as a GA feature when it is implemented, with open questions addressed, including 
+how exactly such CSI driver is installed/delivered to an OCP cluster, and how exactly we want to allow users to
+consume the provided CSI volume.
```

Comment:
We will probably not want to GA this until the related work is also complete.  Without the admission controller+config logic to restrict who can actually use these volumes in a pod, this CSI driver by itself is just a giant security hole.

Reply:
Yeah I was unclear on this based on prior dialogues

I agree and will clarify here.